### PR TITLE
Fixed #5543: implemented "show all opened terminals" quick-open command

### LINKED ISSUE #5543: [quick-open] should have a "show all opened terminals" quick-open command
Similar to #2214, we should have a quick-open command (with a prefix `term`, identical to VS Code) to display all currently opened terminals.

### PR DESCRIPTION
**Pull request by fangnx**
Implemented the "Show All Opened Terminals" command in the quick open menu. The command displays all the currently opened terminal widgets, and the user can navigate to each of them by selection and have the option to create a new terminal.

The command has the global prefix: `term `, to be aligned to VS Code.

The Terminal package may require some refactoring. The `terminal-frontend-contribution` class seems to include functions that should be otherwise put into a separate class.

### CODE REVIEW

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+     * @param {TerminalWidget} widget - the terminal widget to be opened.
+     * @returns Function that would open the terminal if mode === QuickOpenMode.OPEN.
+     */
+    protected getRunFunction(widget: TerminalWidget): (mode: QuickOpenMode) => boolean {
+
+        return (mode: QuickOpenMode) => {
+            if (mode !== QuickOpenMode.OPEN) {
+                return false;
+            }
+            this.shell.activateWidget(widget.id);
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: I believe you can simply call `terminalService.open` at this point and drop the injection of `ApplicationShell`.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+     * @param {TerminalWidget} widget - the terminal widget.
+     * @returns Promise of the quick open item.
+     */
+    protected async toItem(widget: TerminalWidget): Promise<QuickOpenItem<QuickOpenItemOptions>> {
+
+        const options: QuickOpenItemOptions = {
+            label: widget.id,
+            iconClass: widget.title.iconClass,
+            description: widget.title.label,
+            tooltip: '',
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Do you think we should include a `tooltip`? Maybe the `title` will be appropriate.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                enableSeparateSubstringMatching: true
+            }
+        };
+    }
+
+    open(): void {
+        this.prefixQuickOpenService.open(this.prefix);
+    }
+
+    get all(): TerminalWidget[] {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Why are we re-implementing the `all` method?
We already have it.

https://github.com/theia-ide/theia/blob/01993d09d0ade55138db50df4a8b1334180216cb/packages/terminal/src/browser/base/terminal-service.ts#L41

Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: Thanks! I am still working on the PR to adjust to what we just discussed ;)

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Converts the terminal widget to the quick open item.
+     * @param {TerminalWidget} widget - the terminal widget.
+     * @returns Promise of the quick open item.
+     */
+    protected async toItem(widget: TerminalWidget): Promise<QuickOpenItem<QuickOpenItemOptions>> {
+
+        const options: QuickOpenItemOptions = {
+            label: widget.id,
+            iconClass: widget.title.iconClass,
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: For consistency, I don't think the `iconClass` is necessary.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        const widgets: TerminalWidget[] = this.all.sort();
+
+        if (widgets.length === 0) {
+            terminalItems.push(new QuickOpenItem({
+                label: 'List of opened terminals is currently empty',
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Instead of displaying no results, there should be an item to `Create a New Terminal`.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    get all(): TerminalWidget[] {
+        return this.widgetManager.getWidgets(TERMINAL_WIDGET_FACTORY_ID) as TerminalWidget[];
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        const widgets: TerminalWidget[] = this.all.sort();
+
+        if (widgets.length === 0) {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: This item should not be displayed when there are no results.
It should be displayed at all times at the end of the list.

File: packages/terminal/src/browser/terminal-quick-open-service.ts (hunk 5 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: Sure. I am working on implementing it now.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    get all(): TerminalWidget[] {
+        return this.widgetManager.getWidgets(TERMINAL_WIDGET_FACTORY_ID) as TerminalWidget[];
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        const widgets: TerminalWidget[] = this.all.sort();
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: It is not obvious how this method actually sorts.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts [hunk 8]
```diff
@@ -230,6 +251,10 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
             keybinding: 'ctrlcmd+k',
             context: TerminalKeybindingContexts.terminalActive
         });
+        keybindings.registerKeybinding({
+            command: TerminalCommands.SHOW_ALL_OPENED_TERMINALS.id,
+            keybinding: 'ctrlcmd+k ctrlcmd+t'
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts]: I'm not sure this is the correct keybinding.
<kbd>ctrlcmd</kbd>+<kbd>k</kbd> <kbd>ctrlcmd</kbd>+<kbd>t</kbd> is used to change theme in vscode.

File: packages/terminal/src/browser/terminal-frontend-module.ts [hunk 9]
```diff
@@ -103,4 +104,9 @@ export default new ContainerModule(bind => {
 
     bind(TerminalLinkmatcherDiffPost).toSelf().inSingletonScope();
     bind(TerminalContribution).toService(TerminalLinkmatcherDiffPost);
+
+    [CommandContribution, KeybindingContribution, QuickOpenContribution].forEach(serviceIdentifier => {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-frontend-module.ts]: You duplicated the binding found on line 71.

File: packages/terminal/src/browser/terminal-quick-open-service.ts (hunk 7 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: I modified this method to sort widgets by the widget id.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts (hunk 8 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts]: Sure! Just removed the keybinding.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts [hunk 10]
```diff
@@ -28,7 +28,7 @@ import {
 import { QuickPickService } from '@theia/core/lib/common/quick-pick-service';
 import {
     ApplicationShell, KeybindingContribution, KeyCode, Key,
-    KeybindingRegistry, Widget, LabelProvider, WidgetOpenerOptions
+    KeybindingRegistry, Widget, LabelProvider, WidgetOpenerOptions,
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts]: minor, unnecessary change after the refactoring.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    @inject(PrefixQuickOpenService)
+    protected readonly prefixQuickOpenService: PrefixQuickOpenService;
+
+    @inject(WidgetManager)
+    protected readonly widgetManager: WidgetManager;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    // Todo: should be removed once it is refactored.
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Should be `TODO`.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+import { TerminalFrontendContribution, TerminalCommands } from './terminal-frontend-contribution';
+
+@injectable()
+export class TerminalQuickOpenService implements QuickOpenModel, QuickOpenHandler {
+
+    @inject(PrefixQuickOpenService)
+    protected readonly prefixQuickOpenService: PrefixQuickOpenService;
+
+    @inject(WidgetManager)
+    protected readonly widgetManager: WidgetManager;
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Unused injection.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+    open(): void {
+        this.prefixQuickOpenService.open(this.prefix);
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        // Get the list of currently opened terminal widgets, sorted by the widget id.
+        const widgets: TerminalWidget[] = this.terminalService.all
+            .sort((widgetA, widgetB) => (widgetA.id.localeCompare(widgetB.id)));
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: minor, the following would suffice.

```ts
sort((a: TerminalWidget, b: TerminalWidget) => (a.id.localeCompare(b.id)));
```

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        acceptor(terminalItems);
+        return;
+    }
+
+    /**
+     * Get the function that can create a new terminal.
+     * @param {TerminalWidget} widget - the terminal widget to be opened.
+     * @returns Function that would create a new terminal if mode === QuickOpenMode.OPEN.
+     */
+    protected getCreateNewTerminalfunction(): (mode: QuickOpenMode) => boolean {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Please find a better name for this method.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+                return false;
+            }
+            this.terminalService.open(widget);
+            return true;
+        };
+    }
+}
+
+/**
+ * Todo: merge it to TerminalFrontendContribution
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Should be `TODO`.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+            }
+            this.terminalFrontendContribution.openTerminal();
+            return true;
+        };
+    }
+
+    /**
+     * Convert the terminal widget to the quick open item.
+     * @param {TerminalWidget} widget - the terminal widget.
+     * @returns Promise of the quick open item.
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: This `@returns` statement looks off.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts [hunk 17]
```diff
@@ -367,7 +375,7 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
         return ref instanceof TerminalWidget ? ref : undefined;
     }
 
-    protected async openTerminal(options?: ApplicationShell.WidgetOptions): Promise<void> {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts]: I'm not sure about editing the access modifier of the method.
Instead, you can inject the `CommandRegistry` where you call the method, and use the following:

```ts
this.commandRegistry.executeCommand(TerminalCommands.NEW.id);
```

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+     * Get the function that can create a new terminal.
+     * @param {TerminalWidget} widget - the terminal widget to be opened.
+     * @returns Function that would create a new terminal if mode === QuickOpenMode.OPEN.
+     */
+    protected getCreateNewTerminalfunction(): (mode: QuickOpenMode) => boolean {
+        return (mode: QuickOpenMode) => {
+            if (mode !== QuickOpenMode.OPEN) {
+                return false;
+            }
+            this.terminalFrontendContribution.openTerminal();
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: You can do the following [instead](https://github.com/theia-ide/theia/pull/5577/files#r297383606) and not change the access modifier.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts (hunk 17 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts]: Thanks Vince! It works nicely.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+
+    /**
+     * Convert the terminal widget to the quick open item.
+     * @param {TerminalWidget} widget - the terminal widget.
+     * @returns The quick open item.
+     */
+    protected async toItem(widget: TerminalWidget): Promise<QuickOpenItem<QuickOpenItemOptions>> {
+
+        const options: QuickOpenItemOptions = {
+            label: widget.id,
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Are you sure we should display the `id` as the quick-open label? The terminal's `id` contains much more information, so I was just wondering what your thoughts were.

Currently, when running a task, we get the following items in the `term ` quick-open.

<div align='center'>

Screen Shot 2019-06-26 at 8 46 31 AM


</div>

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        const widgets: TerminalWidget[] = this.terminalService.all
+            .sort((a: TerminalWidget, b: TerminalWidget) => (a.id.localeCompare(b.id)));
+
+        for (const widget of widgets) {
+            const item = await this.toItem(widget);
+            terminalItems.push(item);
+        }
+        // Append a quick open item to create a new terminal.
+        const createNewTerminalItem = new QuickOpenItem({
+            label: 'Open New Terminal',
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Do you think we should add an icon class for the create?

```ts
iconClass: 'fa fa-plus'
```

Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: Sure.

File: packages/terminal/src/browser/terminal-quick-open-service.ts (hunk 19 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: I was unsure about what to be displayed. I thought the `title label` (`gitpod/workspace/...` may be too verbose as the label, so I decided to display the `widget id` (terminal-#) as the label and the `title label` as the description.

Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: For comparison, here's vscode:

<div align='center'>

Screen Shot 2019-06-26 at 9 55 29 AM

</div>

I guess in our case I don't see the added value of displaying the `terminal-0` as the first label.
Maybe it's because in our case our `id` is actually `terminal-0` and not simply `0`.

Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: @lmcbout do you have an opinion on this?

**Comment by vince-fugnitto** [comment]
@lmcbout could you test the PR as well?

**Comment by lmcbout** [comment]
Currently testing, Putting terminal-# seems repetitive. I can only see one terminal using  (_terminal 6) , the others, I see the user path, so hard to match the list of terminal (Quick selection) and the actual terminal . I also notice the same number being used for the terminal (See picture) . Also, How is the number generated, if I open a new terminal, (Small number) and if I run a task, (Number 15, 26. , not even consecutive, why?
What do you think about having a separator between the list of terminals and the "Open new terminal"  when there is one or more terminals?

ShowTerminal

**Comment by fangnx** [comment]
> Currently testing, Putting terminal-# seems repetitive. I can only see one terminal using (_terminal 6) , the others, I see the user path, so hard to match the list of terminal (Quick selection) and the actual terminal . I also notice the same number being used for the terminal (See picture) . Also, How is the number generated, if I open a new terminal, (Small number) and if I run a task, (Number 15, 26. , not even consecutive, why?
> What do you think about having a separator between the list of terminals and the "Open new terminal" when there is one or more terminals?
> 
> ShowTerminal

Thanks Jacque! Just found out the sorting logic does not work from your screenshot.

Yes, adding a separator makes sense. The terminal-# is the terminal id. I am not sure why the numbers are not consecutive, but I think since it is difficult to match the terminal-# to the name displayed on the terminal widget, I would use `title id` (`~/theia/.../...`) as a label, just as Vince suggested.

**Comment by fangnx** [comment]
I can't reproduce the repetitive terminal-# issue. How did you get the same terminal-6 when you tested it ;)? @lmcbout

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+            const item = await this.toItem(widget, { groupLabel: '', showBorder: false });
+            terminalItems.push(item);
+        }
+        // Append a quick open item to create a new terminal.
+        const createNewTerminalItem = new QuickOpenGroupItem<QuickOpenGroupItemOptions>({
+            label: 'Open New Terminal',
+            iconClass: 'fa fa-plus',
+            run: this.doCreateNewTerminal(),
+            groupLabel: '',
+            showBorder: true
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: We should only display the label if there currently are terminal-widgets present.

```ts
showBorder: !!terminalItems.length
```

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+        for (const widget of widgets) {
+            const item = await this.toItem(widget, { groupLabel: '', showBorder: false });
+            terminalItems.push(item);
+        }
+        // Append a quick open item to create a new terminal.
+        const createNewTerminalItem = new QuickOpenGroupItem<QuickOpenGroupItemOptions>({
+            label: 'Open New Terminal',
+            iconClass: 'fa fa-plus',
+            run: this.doCreateNewTerminal(),
+            groupLabel: '',
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Should be:

```ts
groupLabel: undefined
```

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        // Get the list of currently opened terminal widgets, sorted by the widget id.
+        const widgets: TerminalWidget[] = this.terminalService.all
+            .sort((a: TerminalWidget, b: TerminalWidget) => (a.title.label.localeCompare(b.title.label)));
+
+        for (const widget of widgets) {
+            const item = await this.toItem(widget, { groupLabel: '', showBorder: false });
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: I don't quite understand why we need to do this, why don't you simply make `toItem` return a `QuickOpenGroupItem` instead off the bat?

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+                enableSeparateSubstringMatching: true
+            }
+        };
+    }
+
+    open(): void {
+        this.prefixQuickOpenService.open(this.prefix);
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Refactor to use:

```ts
items: QuickOpenGroupItem[]
```

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+        const terminalItems: QuickOpenItem[] = [];
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Similar refactoring:

```ts
const terminalItems: QuickOpenGroupItem[] = [];
```

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+            return true;
+        };
+    }
+
+    /**
+     * Convert the terminal widget to the quick open item.
+     * @param {TerminalWidget} widget - the terminal widget.
+     * @returns The quick open group item.
+     */
+    protected async toItem(widget: TerminalWidget, groupOptions: QuickOpenGroupItemOptions): Promise<QuickOpenGroupItem<QuickOpenItemOptions>> {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Should simply return:

```ts
Promise<QuickOpenGroupItem<QuickOpenGroupItemOptions>>
```

Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: No need for `groupOptions`.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+        const options: QuickOpenItemOptions = {
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Should be 
```ts
const options: QuickOpenGroupItemOptions
```

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+            label: widget.title.label,
+            description: widget.id,
+            tooltip: widget.title.label,
+            hidden: false,
+            run: this.getRunFunction(widget)
+        };
+        return new QuickOpenGroupItem<QuickOpenItemOptions>({ ...options, ...groupOptions });
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Drop the `...groupOptions`.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+    open(): void {
+        this.prefixQuickOpenService.open(this.prefix);
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        // Get the list of currently opened terminal widgets, sorted by the widget id.
+        const widgets: TerminalWidget[] = this.terminalService.all
+            .sort((a: TerminalWidget, b: TerminalWidget) => (a.title.label.localeCompare(b.title.label)));
```
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Labels are not unique, do we plan on performing a secondary search when the compare method returns 0 (no change after sort)?

Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: I was thinking about how to sort the terminal widgets. If we sort by the labels, how about doing a secondary sort by the number in widget id (terminal-**1**, terminal-**2**)?

Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: I'm not 100% sure that all terminals will follow the pattern `terminal-{number}` for all created terminals (ex: from extensions, plugins, contributions), so perhaps simply alphabetically as well would suffice?

Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: Ideally you can take out this method and do a custom comparator method so you can have a more complex implementation.

File: packages/terminal/src/browser/terminal-quick-open-service.ts (hunk 21 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: Fixed.

File: packages/terminal/src/browser/terminal-quick-open-service.ts (hunk 22 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: Fixed.

**Comment by vince-fugnitto** [comment]
@lmcbout do you mind reviewing it again when you get the chance?

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+            }
+            this.terminalService.open(widget);
+            return true;
+        };
+    }
+}
+
+/**
+ * TODO: merge it to TerminalFrontendContribution
+ */
```
Comment (lmcbout) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Is it some comments left ?

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+
+import { inject, injectable } from 'inversify';
+import {
+    QuickOpenModel, QuickOpenGroupItem, QuickOpenHandler,
+    QuickOpenOptions, QuickOpenItemOptions, QuickOpenMode,
+    PrefixQuickOpenService,
+    QuickOpenContribution, QuickOpenHandlerRegistry, QuickOpenGroupItemOptions
+} from '@theia/core/lib/browser';
+import { CommandContribution } from '@theia/core/lib/common';
+import { CommandRegistry } from '@theia/core/lib/common';
```
Comment (lmcbout) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: CommandContribution and CommandRegistry are from the same lib, could be put in a single line/
import  {...} ...

Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: @fangnx could you provide this minor fix and I'll merge the PR :)

**Comment by vince-fugnitto** [comment]
@kittaakos could you give it a review as well whenever you get the chance?

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 32]
```diff
@@ -0,0 +1,174 @@
...
+            return true;
+        };
+    }
+}
+
+/**
+ * TODO: merge it to TerminalFrontendContribution
+ */
+@injectable()
+export class TerminalQuickOpenContribution implements CommandContribution, QuickOpenContribution {
```
Comment (akosyakov) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: Why is not it a part of terminal frontend contribution? I cannot get a logic to introduce a new contribution for each command. One extension should have one entry contribution. We need to have something like the design guideline to collect such rules. or add it to the coding guidline.

File: packages/terminal/src/browser/terminal-quick-open-service.ts [hunk 33]
```diff
@@ -0,0 +1,174 @@
...
+import { TerminalCommands } from './terminal-frontend-contribution';
+
+@injectable()
+export class TerminalQuickOpenService implements QuickOpenModel, QuickOpenHandler {
+
+    @inject(PrefixQuickOpenService)
+    protected readonly prefixQuickOpenService: PrefixQuickOpenService;
+
+    @inject(CommandRegistry)
+    protected readonly commandRegistry: CommandRegistry;
```
Comment (akosyakov) [inline packages/terminal/src/browser/terminal-quick-open-service.ts]: to exectue only, you should use `CommandService`

File: packages/terminal/src/browser/terminal-quick-open-service.ts (hunk 32 above)
Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: Doing so would cause the circular dependency issue. The `TerminalQuickOpenService` injects `TerminalService`, which is implemented by `TerminalFrontendContribution`. If `TerminalFrontendContribution` class incluced the code for registering command, it would need to inject the `TerminalQuickOpenService`.

Comment (fangnx) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: > Why is not it a part of terminal frontend contribution? I cannot get a logic to introduce a new contribution for each command. One extension should have one entry contribution. We need to have something like the design guideline to collect such rules. or add it to the coding guidline.

Ideally it should, and putting the coed into a separate class is temporary due to the existing issue.  We are hoping a refactoring (maybe the `TerminalFrontendContribution` should not implement the terminal service) can be done to avoid this type of issues in the Terminal package.

Comment (akosyakov) [reply packages/terminal/src/browser/terminal-quick-open-service.ts]: ok, i see

**Comment by lmcbout** [comment]
While testing with the latest code ( commit 8e42335be0 ), I noticed when selecting a terminal (there is  a duplication of ID - terminal 4), it selects the first one with the "terminal-4", not the selected one which is the second terminal entry (last entry from the list). 

**_Note the last two entries from the terminal list , both have terminal-4_**

WrongTerminalSelected

**Comment by vince-fugnitto** [comment]
> While testing with the latest code ( commit [8e42335](https://github.com/theia-ide/theia/commit/8e42335be0ce1cf7faa36cc8551dd0d758f924c1) ), I noticed when selecting a terminal (there is a duplication of ID - terminal 4), it selects the first one with the "terminal-4", not the selected one which is the second terminal entry (last entry from the list).

I don't believe this is an issue with @fangnx code.
I think it's more our current logic is incorrect when assigning terminal ids. Also, when we lookup
terminals by id it makes sense that it returns the first match anyways, so is this really an issue?

**Comment by vince-fugnitto** [comment]
@lmcbout please open an issue regarding the terminal ids not being unique including steps to reproduce.
